[FIX] never return less than LIMIT results if they exist

hbrunn opened the pull request:
upstream PR: https://github.com/odoo/odoo/pull/32093

hbrunn committed:
[FIX] never return less than LIMIT results if they exist

fixes odoo/#21526

StefanRijnhart reviewed (approved):
Nice one. Tested with ilike operator + explicit ~~ampersands~~ percent characters, and with a table that contains a column called `row_number`. No problems occur. Overhead is acceptable even for queries with large result sets (>100000).

hbrunn commented:
cool, thanks for looking into this! How did you measure? I used this script [1], and get a penalty of roughly 50%, which I consider acceptable because searching usually isn't what costs our time. And the former results are simply wrong.

With this applied:

```shell
testing with 662256 records
         279 function calls (275 primitive calls) in 0.255 seconds

   Ordered by: cumulative time
   List reduced from 119 to 3 due to restriction <3>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.255    0.255 <stdin>:16(<lambda>)
        1    0.000    0.000    0.255    0.255 /home/hbrunn/customers/woonbond/parts/partner-contact/partner_multi_relation/models/res_partner.py:137(search)
        1    0.000    0.000    0.255    0.255 /home/hbrunn/customers/woonbond/parts/odoo/odoo/models.py:1498(search)



         286 function calls (282 primitive calls) in 1.880 seconds

   Ordered by: cumulative time
   List reduced from 123 to 3 due to restriction <3>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    1.880    1.880 <stdin>:17(<lambda>)
        1    0.004    0.004    1.880    1.880 /home/hbrunn/customers/woonbond/parts/partner-contact/partner_multi_relation/models/res_partner.py:137(search)
        1    0.000    0.000    1.876    1.876 /home/hbrunn/customers/woonbond/parts/odoo/odoo/models.py:1498(search)
```

Vanilla:

```shell
testing with 662256 records
         280 function calls (276 primitive calls) in 0.166 seconds

   Ordered by: cumulative time
   List reduced from 119 to 3 due to restriction <3>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.166    0.166 <stdin>:16(<lambda>)
        1    0.000    0.000    0.166    0.166 /home/hbrunn/customers/woonbond/parts/partner-contact/partner_multi_relation/models/res_partner.py:137(search)
        1    0.000    0.000    0.166    0.166 /home/hbrunn/customers/woonbond/parts/odoo/odoo/models.py:1498(search)



         662543 function calls (662539 primitive calls) in 1.208 seconds

   Ordered by: cumulative time
   List reduced from 124 to 3 due to restriction <3>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    1.208    1.208 <stdin>:17(<lambda>)
        1    0.004    0.004    1.208    1.208 /home/hbrunn/customers/woonbond/parts/partner-contact/partner_multi_relation/models/res_partner.py:137(search)
        1    0.000    0.000    1.204    1.204 /home/hbrunn/customers/woonbond/parts/odoo/odoo/models.py:1498(search)
```

[1] script

```python
# cat into odoo-bin shell -d $yourdb
import cProfile, pstats, StringIO

def profile(func):
    pr = cProfile.Profile()
    pr.enable()
    func()
    pr.disable()
    s = StringIO.StringIO()
    ps = pstats.Stats(pr, stream=s).sort_stats('cumulative')
    ps.print_stats(3)
    print s.getvalue()

model = 'res.partner'
print('testing with %d records' % env[model].search([], count=True))
profile(lambda: env[model].search([], count=True))
profile(lambda: env[model].search([]))
```

StefanRijnhart commented:
I mainly checked for larger result sets. The query plan with and without this change is the same except for the additional grouping query. So if I am not mistaken, EXPLAIN ANALYZE tells you the exact overhead. For example, here is the output of a result set of 35 million rows:

```
 Sort  (cost=35064050.15..35152177.05 rows=35250758 width=12) (actual time=1444704.539..1458620.480 rows=37833432 loops=1)
   Sort Key: (min(r.row_number))
   Sort Method: external merge  Disk: 961616kB
   ->  GroupAggregate  (cost=28220985.01..28837873.28 rows=35250758 width=12) (actual time=1367183.630..1404470.571 rows=37833432 loops=1)
         Group Key: r.id
         ->  Sort  (cost=28220985.01..28309111.91 rows=35250758 width=12) (actual time=1367183.615..1382552.616 rows=37833432 loops=1)
               Sort Key: r.id
               Sort Method: external merge  Disk: 961608kB
               ->  Subquery Scan on r  (cost=21025412.30..21994808.14 rows=35250758 width=12) (actual time=1276033.886..1326332.938 rows=37833432 loops=1)
                     ->  WindowAgg  (cost=21025412.30..21642300.56 rows=35250758 width=16) (actual time=1276033.884..1311270.946 rows=37833432 loops=1)
                           ->  Sort  (cost=21025412.30..21113539.19 rows=35250758 width=8) (actual time=1276033.878..1289982.241 rows=37833432 loops=1)
                                 Sort Key: account_move_line.date DESC
                                 Sort Method: external merge  Disk: 665784kB
                                 ->  Bitmap Heap Scan on account_move_line  (cost=794153.95..15160687.42 rows=35250758 width=8) (actual time=50735.933..1234388.649 rows=37833432 loops=1)
                                       Recheck Cond: (journal_id = 10)
                                       Rows Removed by Index Recheck: 158516800
                                       Heap Blocks: exact=15780 lossy=4818638
                                       ->  Bitmap Index Scan on account_move_line_journal_id_index  (cost=0.00..785341.26 rows=35250758 width=0) (actual time=50723.885..50723.885 rows=37839175 loops=1)
                                             Index Cond: (journal_id = 10)
```

Everything under the WindowAgg is equal to the cost without this patch, while everything above is overhead. So the extra time is 1458620ms - 1289982ms = 168638ms, or 13%.

For smaller result sets, the overhead can be of much wider range (up to 100%).

If the incorrect behaviour is only triggered by a (non-left) auto join, would it make sense to do some crude text pattern matching to see if the current query contains such a join before wrapping the query?

hbrunn commented:
ah, thanks.

In my first clumsy version for 7, I added this: https://github.com/hbrunn/OCB/commit/fd97f40d14d561abf224a1fc1a49af0cbff4d3c5

But now I believe only the `auto_join`s of the current model are relevant - if somewhere further down in the path of a proposition's left hand side there's an `auto_join`, it will be this model's `_search`'s task to take care of it, because without `auto_join`, it will be called, so how about

```python
use_complicated_stuff = any(
    self._fields[field_name].auto_join
    for field_name in
    map(
        lambda x: str(x[0]).split('.', 1)[0],
        domain,
    )
    if field_name in self._fields
)
```

? Or better do the same with leaves from the expression object to get expanded domains for function fields, but given we don't have access to this here, we'd have to parse the domain twice and if some function field's domain is expensive to compute, that's not cool. We could stick the expression generated in `_where_calc` into the `Query` object for further reference, seems the best option to me in order not to break APIs. Or just do the analysis in `_where_calc` and set a flag on the query object.

PS: corrected the code several times

StefanRijnhart commented:
No, because as you know I am trying to get Odoo to adopt my patch to make auto-join use a left join by default, and in that case the change of this PR is not needed because we have auto-join enabled for every relational field and we have not encountered the problem that this PR attempts to solve.

However, the full story is that I still need to work on an extension of this patch that fixes a specific combination of related fields on inherited models and what have you, and it seems that a non-left join is sometimes unavoidable.

So I would prefer a match to determine if the query features a non-left join specifically.

StefanRijnhart commented:
Wait, you are right. Make it a property of the Query object if you can. My patch can then be refactored to only set the flag if a non-left join is included in the Query.

hbrunn commented:
but why don't we have a problem with left joins? Didn't want to nitpick about that before, but now it seems relevant. What I'm fixing here is the behavior if one record of the current model has multiple records of the joined model, this will produce duplicate ids in the result set no matter what kind of join we use, right?

(if I wrote something stupid here I'll invoke the new EU copyright directive and not give anyone the license to cite me here)

StefanRijnhart commented:
Good question! But can I then ask you to add a test case that reproduces the problem (did't want to nitpick about that before ;-P) so that I can hold it against my left-join patch.

hbrunn committed:
fixup! [FIX] never return less than LIMIT results if they exist

hbrunn commented:
here we go

hbrunn commented:
and now without the debugger

StefanRijnhart commented:
Thanks! You can forget what I said about left joins then ;-)

```
> odoo/openerp/addons/base/tests/test_expression.py(534)test_auto_join_with_limit()
-> partners = self.env['res.partner'].search(domain, limit=8)
2019-03-27 15:27:42,995 13826 DEBUG odoo openerp.sql_db: query: SELECT "res_partner".id FROM "res_partner" LEFT JOIN "res_users" as "res_partner__user_ids" ON ("res_partner"."id" = "res_partner__user_ids"."partner_id") WHERE ("res_partner__user_ids"."login"::text like '%user%%') ORDER BY "res_partner"."display_name"   limit 8
> odoo/openerp/addons/base/tests/test_expression.py(535)test_00_auto_join_with_limit()
-> self.assertEqual(len(partners), 8)
AssertionError: Assertio...7 != 8',)
```

hbrunn commented:
I'm considering making this a module anyways as I have next to no hope this is going to be merged anywhere. Am I too pessimistic?

StefanRijnhart commented:
Upstream is still very selective about changes like these, but I think this counts as a proper bugfix and we can keep it in OCB while there is still an open PR to upstream.

hbrunn committed:
fixup! [FIX] never return less than LIMIT results if they exist

hbrunn commented:
we actually only need all of this when a limit is involved, so that's then the case where the check for `auto_join` is relevant. Other cases will behave as without the patch, so we shouldn't have any performance penalty when the result wouldn't be potentially wrong otherwise.

It's still hard for me to accept that it's faster to uniquify a list of ints in python while keeping order than have postgres do it, but that seems to be the case.

Ready for review I think.

StefanRijnhart reviewed (approved):
Excellent! I love it how you solved the problem while preventing the overhead where possible.
